Add ExternalDeploymentTriggerDagRunOperator

odaneau-astro opened the pull request:
Adds an ExternalDeploymentTriggerDagRunOperator to start new Dag runs on external Airflow deployments through the Airflow REST API.

rajaths010494 commented:
We need to defer and trigger the dag from Trigger not in execute method.
Adding a reference for the same sensor and how it is implemented https://github.com/astronomer/astronomer-providers/blob/main/astronomer/providers/core/sensors/external_task.py#L38-L62.

phanikumv reviewed (commented):
@odaneau-astro please let us know if you need any help to move this forward

odaneau-astro commented:
Thanks for the offer. Time was missing over the last few weeks. 
The code is mostly done, I'm adding tests and will be able to open the PR sometime this week.

odaneau-astro commented:
I have considered using the Airflow API Python client at https://github.com/apache/airflow-client-python
Chose to build a custom Client instead since not all endpoints are currently supported.
Curious to have feedback on this.

pankajkoti commented on astronomer/providers/core/example_dags/example_external_deployment_trigger_dag_operator.py:
```diff
@@ -0,0 +1,82 @@
+)
+
+DEPLOYMENT_CONN_ID = os.getenv("ASTRO_DEPLOYMENT_CONN_ID", "")
+ASTRO_AUTH_CONN_ID = os.getenv("ASTRO_AUTH_CONN_ID", "")
```
If we have one, can we please add a link to the documentation in a comment here describing how to setup an Astro auth connection?

pankajkoti commented on astronomer/providers/core/example_dags/example_external_deployment_trigger_dag_operator.py:
```diff
@@ -0,0 +1,82 @@
+    schedule_interval=None,
+    catchup=False,
+    default_args=default_args,
+    tags=["example", "async", "http"],
```
Do we need an `http` tag here?

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+
+class ExternalDeploymentTriggerDagRunOperator(BaseOperator):
+    """
+        External Deployment Trigger Dag Run Operator Make HTTP call to trigger a Dag Run and poll for the response state of externally
+        deployed DAG run to complete. The host should be external deployment url, header must contain access token
```
```suggestion
        External Deployment Trigger Dag Run Operator makes an HTTP call to trigger a run of an externally deployed Dag and poll for its status asynchronously. The host should be the external deployment URL and the header must contain a valid access token
```

pankajkoti replied:
Might need to check the indentation afterwards if you consider the suggestion :)

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+        :param allowed_states: List of allowed states, default is ``['success']``.
+        :param failed_states: List of failed or dis-allowed states, default is ``None``.
+
+    Changed in version 2.1.3: 'queued' is added as a possible value.
```
possible value for what (do we mean here)?

odaneau-astro replied:
Removed it. It was copy-pasted by mistake.

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+        "conf",
+    )
+    template_fields_renderers = {"headers": "json", "data": "py", "conf": "py"}
+    ui_color = "#7352ba"
```
Will it be possible to add a comment here regarding what colour this hash code corresponds to :)?
Looks like `Blue Marguerite` https://colors.dopely.top/color-pedia/7352ba

Also curious to understand if there was some significance behind choosing this colour. Is it that we use it in some other places too?

odaneau-astro replied:
removed it completely. no need to be fancy with custom UI colors, the default one works.

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+        :param poke_interval: Poke interval to check dag run status when wait_for_completion=True.
+            (default: 60)
+        :param allowed_states: List of allowed states, default is ``['success']``.
+        :param failed_states: List of failed or dis-allowed states, default is ``None``.
```
Isn't the default state `[State.FAILED]`?

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+        """
+        execution_date = parse_execution_date(self.execution_date)
+
+        api = AirflowApiClient(
```
```suggestion
        api_client = AirflowApiClient(
```

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+
+        if dag_run is None:
+            raise RuntimeError("The dag_run should be set here!")
+        logging.warning(type(dag_run))
+        logging.warning(inspect.getmembers(dag_run))
```
Why are we setting the log level to `warning` here?

pankajkoti commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,417 @@
+
+    :param json_dict: Dictionary containing DagRun properties to parse
+    """
+    if (data_interval_start := json_dict.get("data_interval_start")) and (
```
Sadly, the [setup.cfg](https://github.com/astronomer/astronomer-providers/blob/main/setup.cfg#L32) mentions `python>=3.7`. So the usage of walrus operators may not be supported on `3.7`

If we wish to continue using this, we might need to raise an exception while using the operator is Python version `<3.8` or refactor this bit here to also support `3.7 `

pankajkoti commented on astronomer/providers/core/triggers/external_dagrun.py:
This file name seems to be inconsistent with the operator file name. In general, we try to keep it consistent.

pankajkoti commented on tests/core/operators/test_external_dagrun.py:
```diff
@@ -0,0 +1,409 @@
+#             http_conn_id=self.CONN_ID,
+#             trigger_dag_id=self.TRIGGER_DAG_ID,
+#             trigger_run_id=self.TRIGGER_RUN_ID,
+#             conf=self.CONF,
+#             execution_date=datetime(2023, 1, 1),
+#             unpause_dag=True,
+#             reset_dag_run=True,
+#             wait_for_completion=True,
+#             headers=self.HEADERS,
+#             poke_interval=5,
+#         )
+#
+#         with mock.patch.object(operator.log, "info") as mock_log_info:
+#             operator.execute_complete(context={}, event={"state": mock_state})
+#         mock_log_info.assert_called_with(mock_message, {"state": mock_state})
+#
+#     @pytest.mark.parametrize(
+#         "mock_resp, mock_message",
+#         [
+#             ({"state": "error"}, "Task Failed with response: %s"),
+#             ({"test": "test"}, "Task Failed with response: %s"),
+#         ],
+#     )
+#     def test_external_deployment_execute_complete_value_error(self, mock_resp, mock_message):
+#         """Assert execute_complete method to raise Value error"""
+#         operator = ExternalDeploymentTriggerDagRunOperator(
+#             task_id="external_deployment_trigger_dag_run",
+#             http_conn_id=self.CONN_ID,
+#             trigger_dag_id=self.TRIGGER_DAG_ID,
+#             trigger_run_id=self.TRIGGER_RUN_ID,
+#             conf=self.CONF,
+#             execution_date=datetime(2023, 1, 1),
+#             unpause_dag=True,
+#             reset_dag_run=True,
+#             wait_for_completion=True,
+#             headers=self.HEADERS,
+#             poke_interval=5,
+#         )
+#         with pytest.raises(ValueError):
+#             operator.execute_complete(context={}, event=mock_resp)
```
Do we not need this block? If not, can we remove the commented block?

pankajkoti reviewed (commented):
Looks great efforts to me 👏🏽 . Have some comments inline. Also, did we get a chance to test the operator and trigger such an externally deployed DAG to validate the operator end to end?

odaneau-astro commented:
I'm sorry I let this PR open for so long. Was working on PS engagements and forgot about it.
I'll address all the comments and fix the failed CI checks this week.

odaneau-astro commented:
@phanikumv : Fixed all the errors and comments.
Somehow mypy still throws an error on  astronomer/providers/microsoft/azure/hooks/wasb.py:59.
It is not something I've changed with this PR. Let me know what you want to do with it.

pankajastro commented on astronomer/providers/core/operators/external_trigger_dagrun.py:
```diff
@@ -0,0 +1,412 @@
+
+    def get_link(self, operator: BaseOperator, *, ti_key: TaskInstanceKey) -> str:
+        """Provide link to browse externally triggered Dag run."""
+        # execution_date = XCom.get_value(ti_key=ti_key, key=XCOM_EXECUTION_DATE_ISO)
```
do we need this or we can remove?

pankajastro commented:
Look like maybe some more tests require.

phanikumv commented:
@odaneau-astro this PR is going on since quite some time. not sure whether we still need it?

phanikumv commented:
Closing as there is no activity on this since 2 months
